### PR TITLE
ENH : add PDF Opening Destination property

### LINKED ISSUE #1425: JS Executing Twice
Hi 

I'm attempting to insert JS that opens the URL whenever the user opens the PDF. Whatever JS I add, it executes it twice. In my scenario, Adobe Acrobat is alerting me about security twice because the same JS was executed twice.

## Environment

Plaform: `Linux-4.15.0-108-generic-x86_64-with-Ubuntu-18.04-bionic`
PyPDF2 Version: `1.26.0`
Python : 3.6.9

## Code + PDF

```
from PyPDF2 import PdfFileWriter, PdfFileReader

output = PdfFileWriter()
ipdf = PdfFileReader(open('test.pdf', 'rb'))

for i in range(ipdf.getNumPages()):
	page = ipdf.getPage(i)
	output.addPage(page)

with open('updated_pdf.pdf', 'wb') as f:
	output.addJS('app.launchURL("https://google.com/", true);')
	output.write(f)

```
[test.pdf](https://github.com/py-pdf/PyPDF2/files/9952765/test.pdf)
[updated_pdf.pdf](https://github.com/py-pdf/PyPDF2/files/9952685/updated_pdf.pdf)

### Todo:
JS should be executed only once

### PR DESCRIPTION
**Pull request by pubpub-zz**
(fix JS executed twice extracted in another PR)
add capability to define an opening destination for the PDF

### CODE REVIEW

**Comment by MartinThoma** [comment]
@pubpub-zz Is the new `opening` property related in any way to the issues you mentioned?

**Comment by MartinThoma** [comment]
In general I prefer PRs which deal with a single issue. This makes it way easier for me to merge a PR as I only have to check one thing at a time.

**Comment by pubpub-zz** [comment]
> In general I prefer PRs which deal with a single issue. This makes it way easier for me to merge a PR as I only have to check one thing at a time.

I've introduced the `opening` at the same time as the  change was removing access to AutoOpen field

**Comment by MartinThoma** [comment]
I think splitting would be nice :pray: 

I'm uncertain about the naming `opening`, but the JS-fixes are very clearly something I want to merge

**Comment by pubpub-zz** [comment]
> I think splitting would be nice 🙏
> 
> I'm uncertain about the naming `opening`, but the JS-fixes are very clearly something I want to merge

any ideas another name for opening ?

**Comment by MartinThoma** [comment]
If you want to make `OpenAction` accessible, I would just do that. But there is more going on in the code.

According to `TABLE 3.25 Entries in the catalog dictionary` the `OpenAction` entry is either an array or a dictionary:

> (Optional; PDF 1.1) A value specifying a destination to be displayed or
an action to be performed when the document is opened. The value is
either **an array defining a destination** (see Section 8.2.1, “Destinations”)
or **an action dictionary representing an action** (Section 8.5, “Actions”). If
this entry is absent, the document should be opened to the top of the
first page at the default magnification factor.

Why is the `isinstance(oa, (str, bytes))` helpful?

I'm sorry if I ask dumb questions; I know way less about PDF than you :sweat_smile:

**Comment by pubpub-zz** [comment]
> If you want to make `OpenAction` accessible, I would just do that and call it `open_action`. But there is more going on in the code.

I dislike the idea of calling it open_action as actually the JS actions are now executed from other Names/JavaScript. 
OpenAction will now only deal with destinations to define the opening of the pdf

> 
> According to `TABLE 3.25 Entries in the catalog dictionary` the `OpenAction` entry is either an array or a dictionary:
> 
> > (Optional; PDF 1.1) A value specifying a destination to be displayed or
> > an action to be performed when the document is opened. The value is
> > either **an array defining a destination** (see Section 8.2.1, “Destinations”)
> > or **an action dictionary representing an action** (Section 8.5, “Actions”). If
> > this entry is absent, the document should be opened to the top of the
> > first page at the default magnification factor.
> 
> Why is the `isinstance(oa, (str, bytes))` helpful?

You can define named destinations. Even if not in the PDF standard it works (I was also doubtfull about it😉). For the getter I've prefered to keep the name instead of the actual destination for users to not believe that something has been modified


I've extracted the fix on  in another PR

**Comment by pubpub-zz** [comment]
Waiting for #1439 to be merged before updating this one

**Comment by pubpub-zz** [comment]
@MartinThoma 
looks ready

**Comment by MartinThoma** [comment]
@MasterOdin What do you think about newly added `opening` property?

File: PyPDF2/_writer.py [hunk 1]
```diff
@@ -411,6 +411,54 @@ def insertBlankPage(
...
+        property setting the opening Destination or Page or NamedDest (`str`)
+        None:  removes the opening entry
+        :param destination:.
+
+        (value stored in "/OpenAction" entry in the Pdf Catalog)
+        """
+        if dest is None:
+            try:
+                del self._root_object["/OpenAction"]
+            except Exception:
```
Comment (MasterOdin) [inline PyPDF2/_writer.py:449]: I'm not a fan of the bare `Exception` catch as failing to delete for something other than a `KeyError` strikes me as violating the user expectation of what this function does.

I'd say to only catch on `KeyError` and let all others bubble up:
```suggestion
            except KeyError:
```

File: tests/test_javascript.py [hunk 2]
```diff
@@ -47,3 +48,41 @@ def get_javascript_name():
     assert (
         first_js != second_js
     ), "add_js should add to the previous script in the catalog."
+
+
+def test_startup_dest(pdf_file_writer):
```
Comment (MasterOdin) [inline tests/test_javascript.py:53]: The `/OpenAction` property doesn't explicitly have to do with JS, so I'd vote to move this into `test_writer.py`.

File: PyPDF2/_writer.py [hunk 3]
```diff
@@ -411,6 +411,54 @@ def insertBlankPage(
         deprecate_with_replacement("insertBlankPage", "insert_blank_page")
         return self.insert_blank_page(width, height, index)
 
+    @property
+    def opening(self) -> Union[None, Destination, TextStringObject, ByteStringObject]:
+        """
+        property returning the opening Destination
+        returns None if no destination is set
+
+        (value stored in "/OpenAction" entry in the Pdf Catalog)
```
Comment (MasterOdin) [inline PyPDF2/_writer.py:420]: Maybe something like:
```suggestion
        Property returning the opening destination ("/OpenAction" entry in the
        PDF catalog). Set to `None` to default to opening at top of first page.
```

File: PyPDF2/_writer.py [hunk 4]
```diff
@@ -411,6 +411,54 @@ def insertBlankPage(
...
+            return None
+
+    @opening.setter
+    def opening(self, dest: Union[None, str, Destination, PageObject]) -> None:
+        """
+        property setting the opening Destination or Page or NamedDest (`str`)
+        None:  removes the opening entry
+        :param destination:.
+
+        (value stored in "/OpenAction" entry in the Pdf Catalog)
```
Comment (MasterOdin) [inline PyPDF2/_writer.py:444]: I'm not sure need to set this for sphinx to pick up on it? Seems like the setter docstring is omitted in other places.

**Comment by MasterOdin** [comment]
Personally, I feel like calling it `open_action` would be better as `opening` doesn't feel right as a noun saying "PDF Writer opening". I could see calling it maybe `opening_action` as in "PDF Writer opening action" vs `open_action`.

I do think it's fine to add this property, even if it's not really one that's often used given it is in the PDF spec (see Table 28 in 7.7.2).

**Comment by pubpub-zz** [comment]
> Personally, I feel like calling it `open_action` would be better as `opening` doesn't feel right as a noun saying "PDF Writer opening". I could see calling it maybe `opening_action` as in "PDF Writer opening action" vs `open_action`.
> 
> I do think it's fine to add this property, even if it's not really one that's often used given it is in the PDF spec (see Table 28 in 7.7.2).

The point is that this property is not dealing with any action/script. What about `open_destination` ?

**Comment by MasterOdin** [comment]
`open_destination` makes sense.

File: PyPDF2/_writer.py (hunk 4 above)
Comment (pubpub-zz) [reply PyPDF2/_writer.py:442]: merged

**Comment by pubpub-zz** [comment]
@MartinThoma 
it should be good now for you

**Comment by MartinThoma** [comment]
Thank you @pubpub-zz for putting that much work again in this PR :heart: 

And thank you @MasterOdin for the thorough review :hugs:

**Pull request merged**